packages/airnode-node/src/handlers/call-api.ts

```diff
@@ -81,28 +81,79 @@ export interface CallApiPayload {
   readonly aggregatedApiCall: AggregatedApiCall;
 }
 
-interface VerificationFailure {
-  log: PendingLog;
-  response: ApiCallErrorResponse;
-}
-
-function verifySponsorWallet(payload: CallApiPayload): VerificationFailure | null {
+function verifySponsorWallet(payload: CallApiPayload): LogsData<ApiCallErrorResponse> | null {
   const { config, aggregatedApiCall } = payload;
   if (aggregatedApiCall.type === 'testing-gateway') return null;
 
   const { sponsorAddress, sponsorWalletAddress, id } = aggregatedApiCall;
   const hdNode = getMasterHDNode(config);
   if (isValidSponsorWallet(hdNode, sponsorAddress, sponsorWalletAddress)) return null;
 
+  // TODO: Abstract this to a logging utils file
   const message = `${RequestErrorMessage.SponsorWalletInvalid}, Request ID:${id}`;
   const log = logger.pend('ERROR', message);
-  return {
-    log,
-    response: {
+  return [
+    [log],
+    {
+      success: false,
+      errorMessage: message,
+    },
+  ];
+}
+
+function verifyRequestId(payload: CallApiPayload): LogsData<ApiCallErrorResponse> | null {
+  const { aggregatedApiCall } = payload;
+  if (aggregatedApiCall.type === 'testing-gateway') return null;
+
+  if (isValidRequestId(aggregatedApiCall)) return null;
+
+  const message = `${RequestErrorMessage.RequestIdInvalid}. Request ID:${aggregatedApiCall.id}`;
+  const log = logger.pend('ERROR', message);
```

Comment:
Logging at the call place is better then propagating the error and logging elsewhere since it's easier to debug (you just find out the error string and you are in the correct place in the code which caused the failure).

However, we return a response as well which is needed for the output of testing gateway.

Reply:
Yeah I also don't like propagating the logs (especially across worker functions), but it was a trade-off to not have to pass in the entire context to every function

Reply:
We actually mean a different thing (I think). I think the pending logs are reasonable and they are still logged at the call site (they are just printed by some other function). What I dislike is this pattern:

```js
const fnA = () => {
try{
const x = someFnCall()
...
return x
} catch (e) {
return null
}
}

const fnB = () => {
const res = fnA()
// The error happens in fnA, but the message is printed by fnB
if (!res) logger.error("An error happened...")
}
```

Reply:
Ah right, my mistake. Yeah I agree with you